packages/cosmic-swingset/src/chain-main.js

```diff
@@ -607,74 +613,81 @@ export default async function main(progname, args, { env, homedir, agcc }) {
     }
   }
 
-  /** @type {Awaited<ReturnType<typeof launch>>['blockingSend'] | undefined} */
-  let blockingSend;
-
   async function toSwingSet(action, _replier) {
     // console.log(`toSwingSet`, action);
-    if (action.vibcPort) {
-      portNums.dibc = action.vibcPort;
-    }
 
-    if (action.storagePort) {
-      // Initialize the storage for this particular transaction.
-      // console.log(` setting portNums.storage to`, action.storagePort);
-      portNums.storage = action.storagePort;
-    }
+    await null;
 
-    if (action.vbankPort) {
-      portNums.bank = action.vbankPort;
-    }
+    switch (action.type) {
+      case AG_COSMOS_INIT: {
+        // console.error('got AG_COSMOS_INIT', action);
 
-    if (action.lienPort) {
-      portNums.lien = action.lienPort;
-    }
+        !blockingSend || Fail`Swingset already initialized`;
 
-    // Snapshot actions are specific to cosmos chains and handled here
-    if (action.type === ActionType.COSMOS_SNAPSHOT) {
-      const { blockHeight, request, args: requestArgs } = action;
-      writeSlogObject?.({
-        type: 'cosmic-swingset-snapshot-start',
-        blockHeight,
-        request,
-        args: requestArgs,
-      });
+        if (action.vibcPort) {
+          portNums.dibc = action.vibcPort;
+        }
 
-      const resultP = handleCosmosSnapshot(blockHeight, request, requestArgs);
+        if (action.storagePort) {
+          portNums.storage = action.storagePort;
+        }
 
-      resultP.then(
-        result => {
-          writeSlogObject?.({
-            type: 'cosmic-swingset-snapshot-finish',
-            blockHeight,
-            request,
-            args: requestArgs,
-            result,
-          });
-        },
-        error => {
-          writeSlogObject?.({
-            type: 'cosmic-swingset-snapshot-finish',
-            blockHeight,
-            request,
-            args: requestArgs,
-            error,
-          });
-        },
-      );
+        if (action.vbankPort) {
+          portNums.bank = action.vbankPort;
+        }
 
-      return resultP;
-    }
+        if (action.lienPort) {
+          portNums.lien = action.lienPort;
+        }
```

Comment:
Consider hardening `portNums` here if this is the only place we initialize it:
```suggestion
        }
        harden(portNums);
```